### PR TITLE
Feature: subscript JSON using a string key path

### PR DESCRIPTION
**Pull request by cjmconie**
This PR adds the ability to query into a JSON type using a key path.

### CODE REVIEW

**Comment by cjmconie** [comment]
I closed this by accident – am reopening now.

File: GenericJSON/Querying.swift [hunk 1]
```diff
@@ -74,4 +74,33 @@ public extension JSON {
     subscript(dynamicMember member: String) -> JSON? {
         return self[member]
     }
+    
+    /// Return the JSON type at the keypath if this is an `.object`, otherwise `nil`
+    ///
+    /// This lets you write `json.valueForKeyPath("foo.bar.jar")`.
```
Comment (zoul) [inline GenericJSON/Querying.swift]:
```suggestion
    /// This lets you write `json[keyPath: "foo.bar.jar"]`.
```

Also, it would be good to give examples, but we can easily do that later after merging.

File: GenericJSONTests/QueryingTests.swift [hunk 2]
```diff
@@ -58,4 +58,27 @@ class QueryingTests: XCTestCase {
...
+                "str": "col",
+                "arr": [1, 2, 3],
+                "obj": [
+                    "x": "rah",
+                    "y": "tar",
+                    "z": "yaz"
+                ]
+            ]
+        ]
+        XCTAssertEqual(json[keyPath: "string"], JSON.string("foo bar"))
```
Comment (zoul) [inline GenericJSONTests/QueryingTests.swift]: This can be simplified to:

```swift
XCTAssertEqual(json[keyPath: "string"], "foo bar")
XCTAssertEqual(json[keyPath: "boolean"], true)
XCTAssertEqual(json[keyPath: "number"], 123)
XCTAssertEqual(json[keyPath: "object.str"], "col")
XCTAssertEqual(json[keyPath: "object.arr"], [1, 2, 3])
XCTAssertEqual(json[keyPath: "object.obj.y"], "tar")
```

This is enabled by the `ExpressibleByXXXLiteral` that makes it possible for the compiler to “automatically upgrade” a plain value to `JSON` when an instance of `JSON` is expected. Here, the compiler can see that the second argument to `XCTAssertEqual` has to be of the same type as the first one, and since the first one is a `JSON?` (as we are calling `JSON.subscript(keyPath:)`), the second is automatically converted to `JSON`, too.

In other words, conforming `JSON` to the `ExpressibleByXXXLiteral` protocols makes it possible for plain value literals such as `"foo"`, `true` or `42` to express JSON values without any extra fluff. Which kind of makes sense and is nice.

**Review by zoul — COMMENTED** [review]
Commit fe17115 should be left out of the history. The file just contains your Xcode editor state and should be kept out of the repository. (I even have a `xcuserdata/` pattern in my global `~/.gitignore` so that I never commit the file accidentally.)

**Review by zoul — CHANGES_REQUESTED** [review]
I think we could rewrite `subscript(keyPath:)` using a helper function to avoid splitting and joining the key path repeatedly, but we can do that later.

**Comment by zoul** [comment]
This is how the helper function could work:

```swift
public subscript(keyPath keyPath: String) -> JSON? {
    return queryKeyPath(keyPath.components(separatedBy: "."))
}

func queryKeyPath<T>(_ path: T) -> JSON? where T: Collection, T.Element == String {

    // Only object values may be subscripted
    guard case .object(let object) = self else {
        return nil
    }

    // Is the path non-empty?
    guard let head = path.first else {
        return nil
    }

    // Do we have a value at the required key?
    guard let value = object[head] else {
        return nil
    }

    let tail = path.dropFirst()

    return tail.isEmpty ? value : value.queryKeyPath(tail)
}
```

The `path` argument to `queryKeyPath` can’t be a simple `[String]`, since `path.dropFirst()` then produces `ArraySlice<String>` and we would not be able to call `queryKeyPath` recursively. This way we only split the path string once and then just juggle the indices, which should be super-effective. What do you think?

File: GenericJSONTests/QueryingTests.swift (hunk 2 above)
Comment (cjmconie) [reply GenericJSONTests/QueryingTests.swift]: Thanks, I wasn't aware of the automatic upgrade the compiler does. I have updated the tests.

File: GenericJSON/Querying.swift (hunk 1 above)
Comment (cjmconie) [reply GenericJSON/Querying.swift]: Updated.

**Comment by cjmconie** [comment]
I have made the changes you requested (including your key path implementation).

Regarding the `queryKeyPath<T>(:T)` helper. Would creating a standard array (using the `init<S>(S)`) from the `ArraySlice<String>` have solved the recursion issue? If so, I guess it would still impose a performance penalty having to init a new array each time? But that is a matter of interest only. Your proposed solution is great.

**Comment by zoul** [comment]
> Regarding the `queryKeyPath<T>(:T)` helper. Would creating a standard array from the `ArraySlice<String>` have solved the recursion issue? If so, I guess it would still impose a performance penalty having to init a new array each time?

Yes, that’s exactly the way I understand it from reading [Strings in Swift 4](https://oleb.net/blog/2017/11/swift-4-strings/) by Ole Begemann. Strings are a special case of the same principle: Creating a substring by _copying_ a part of the underlying storage would be much slower than just keeping track of different start and end indices in the same storage. But if we create a substring by keeping a reference to the underlying storage and just changing the indices, the short substring could potentially keep a huge string storage in memory by referencing it. We solve the problem using a separate type (`ArraySlice` or `Substring`) for intermediate computations. The slices refer to the same underlying storage, keeping it in memory, but the intermediate computations can be very fast and we can allocate a “real” string to return when we’re done. (I hope this is comprehensible and not wrong.)

Thank you for the changes! Will merge now. 🎉

**Review by zoul — APPROVED** [review]
